src/main.c

```diff
@@ -603,12 +656,12 @@ poe_reply_consume(unsigned char *reply)
 	free(cmd);
 
 	if ((reply[0] != cmd_id) || (reply[0] > ARRAY_SIZE(reply_handler))) {
-		ULOG_DBG("received reply with bad command id\n");
+		ULOG_INFO("received reply with bad command id\n");
```

Comment:
Hmm, what's the reasoning behind escalating this from DBG to INFO?

Reply:
This was during debugging, I wanted to be sure that I see if something really bad happens during transfers. But this is not necessary, I shall fix this.
![Screenshot from 2022-07-27 20-51-33](https://user-images.githubusercontent.com/62717533/181351935-95dae24a-2ab4-40f1-8bf4-658ca6d351f8.png)

The I2C protocol produces a large number of failed transfers, about 1% fail. Compared to the OEM I2C-GPIO driver (which has optimizations specific to the RTL SoCs), the timing of the transfers look horrible, see the above image. In the above example, the reply checksum fails, probably because the command was not understood by the MCU, and it would be sufficient to retry. I am just now wondering how this can be done better in `poe_reply_consume()`:

```
	cmd = list_first_entry(&cmd_pending, struct cmd, list);
	list_del(&cmd->list);
	cmd_id = cmd->cmd[0];
	cmd_seq = cmd->cmd[1];

	for (i = 0; i < 11; i++)
		sum += reply[i];

	if (reply[11] != sum) {
		ULOG_DBG("received reply with bad checksum\n");
		free(cmd);
		return -1;
	}
```
The command needs to be resent and not discarded when the checksumming fails. If commands during the startup fail, then PoE fails to work.

Reply:
There's some logic for this being prepared on the UART side, see #15. Since the logic has been developed for the other medium, it'd be quite nice if they could share code-paths completely.

Reply:
@bkobl, that transfer looks way faster than the capture you shared on the forum. Do the I2C lines go over opto-isolators, or isolation transformers? If that's the case, you might get signal integrity issues if you drive the bus too fast. It looks like the MCU is trying to do clock stretching to slow down the transfer.

Also, reply ff means `not-ready`. That also happens on UART when sending requests in fast succession. I could be the same root cause as #10 .

Reply:
I investigated different clock speeds a bit more. Slowing down the bus to 10kHz isn't making anything better. Instead it is becoming obvious that there is an issue with the timing of the I2C bitbanging
![Screenshot from 2022-08-01 08-59-41](https://user-images.githubusercontent.com/62717533/182098914-8ab57069-4792-4273-a973-033d118cb9cd.png)

It looks like every 2ms there is a pause, or maybe the clock is slowed down. This is an RTL839x system with a second VPE and a second clock which is used as a timer (i.e. it can be used to make the CPU sleep and then have an IRQ triggered to wake it up, including switching the VPE context at that point). This is much more complex than the situation on the RTL838x. Althought the transfers look horrible in the trace, they actually succeed. I don't think there is clock slowing on the side of the MCU. This is usually used when there is an issue with the I2C peripherial, not to indicate that the MCU has not finished computations.

The problem with the all FF response seems indeed to come from the MCU not being ready, it seems to happen within a rapid sequency of commands, not asking for monitor data. So I'll do some retrying, you said there was already some logic.